core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -654,85 +650,86 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
-                // start inlining while keeping track of all sequences to check for recursion
-                resolvedSequence flatMap { sequence =>
-                    inlineComponentsAndCountAtomicActions(0, resComponents.toList, List(sequence))
+            // resolve the action document id (if it's in a package/binding);
+            // this assumes that entityStore is the same for actions and packages
+            WhiskAction.resolveAction(entityStore, sequenceAction) flatMap { action =>
+                val resolvedComponents = components map { c => WhiskAction.resolveAction(entityStore, c) }
+                Future.sequence(resolvedComponents) flatMap { sequence =>
+                    // start inlining while keeping track of all sequences to check for recursion
+                    inlineComponentsAndCountAtomicActions(0, sequence.toList, List(action))
                 }
             }
         }
 
         future recoverWith {
-            case _: TooManyActionsInSequence =>
-                Future failed RejectRequest(BadRequest, "too many actions in sequence")
-            case _: SequenceWithRecursion =>
-                Future failed RejectRequest(BadRequest, "recursion detected in sequence")
-            case _: NoDocumentException =>
-                Future failed RejectRequest(BadRequest, "action not found")
+            case _: TooManyActionsInSequence => Future failed RejectRequest(BadRequest, sequenceIsTooLong)
+            case _: SequenceWithRecursion    => Future failed RejectRequest(BadRequest, sequenceIsCyclic)
+            case _: NoDocumentException      => Future failed RejectRequest(BadRequest, sequenceComponentNotFound)
         }
     }
 
-    /*
-     * helper function that traverses the actions to inline and counts atomic actions while keeping track of sequences
+    /**
+     * Traverses the components that make up a sequence to inline any that are sequences.
+     * Counts atomic (non-sequence) actions while keeping track of sequences.
+     *
      * @param atomicActionsCnt the atomic actions seen so far
      * @param actionsToInline the actions to inline
-     * @param sequences the sequences encountered during inlining
-     * @return a successful future in case the conditions were met or a failed fugure with either too many actions or recursion found
+     * @param visitedSequences the sequences already visited during recursive traversal
+     * @return a successful future in case the conditions were met or a failed future with an appropriate error response
      */
-    private def inlineComponentsAndCountAtomicActions(atomicActionsCnt: Int, actionsToInline: List[FullyQualifiedEntityName], sequences: List[FullyQualifiedEntityName])(
-        implicit transid: TransactionId): Future[Boolean] = {
-        if (atomicActionsCnt > actionSequenceLimit)
-            Future.failed(new TooManyActionsInSequence())
-        else {
+    private def inlineComponentsAndCountAtomicActions(
```

Comment:
This method does not implement/enforce the desired invariant: that the sequence is not cyclic. I found the use of "recursive" to be misleading in that regard and renamed variable/updated comments accordingly.

The implementation is too restrictive and exclude the same sequence from appearing twice in the component list of another sequence.